vo_tct: write frame fully instead of every pixel

kasper93 opened the pull request:
This is multiple times faster than just writing every pixel sequence separately. Especially on slower terminal emulators. In general no need to stress I/O, while we can just prepare the frame to print and do it once.

Traneptora commented on video/out/vo_tct.c:
```diff
@@ -101,37 +102,24 @@ static int rgb_to_x256(uint8_t r, uint8_t g, uint8_t b)
-#else
-    printf("%s;%d;%d;%dm", prefix, (int)r, (int)g, (int)b);
-#endif
+    bstr_xappend(NULL, frame, prefix);
```
bstr_xappend calls a realloc and a memcpy. Do we want to do this five times, or would it make more sense to do a bstr_asprintf?

kasper93 replied:
There is already pregenerated lut for those values, specifically to avoid constant asprintf, which in itself involves some overhead.

realloc will happen only on first frame, the buffer is reused. memcpy will happen, but memory copy is the most basic operation we can do, asprintf would do a lot more.

Traneptora commented on video/out/vo_tct.c:
```diff
@@ -101,37 +102,24 @@ static int rgb_to_x256(uint8_t r, uint8_t g, uint8_t b)
-#endif
+    bstr_xappend(NULL, frame, prefix);
+    bstr_xappend(NULL, frame, (bstr){ lut[c].str, lut[c].width });
+    bstr_xappend(NULL, frame, bstr0("m"));
```
Same comment as earlier wrt malloc and memcpy.

kasper93 replied:
See my above answer.

Traneptora commented on video/out/vo_tct.c:
```diff
@@ -101,37 +102,24 @@ static int rgb_to_x256(uint8_t r, uint8_t g, uint8_t b)
 }
 
-static void print_seq3(struct lut_item *lut, const char* prefix,
+static void print_seq3(bstr *frame, struct lut_item *lut, bstr prefix,
```
It seems we have some inconsistency here with regard to using `bstr` vs `struct bstr`. While one is `typedef`-ed to the other, I'm not sure it makes sense to be inconsistent. I recognize that the codebase itself is inconsistent in other places, but maybe we should pick one convention and stick to it?

kasper93 replied:
We should use `bstr` everywhere, I used `struct bstr` in priv, because it was nicely aligned with all other struct there, I can revert that.

Traneptora commented on video/out/vo_tct.c:
```diff
@@ -180,17 +167,16 @@ static void write_half_blocks(
+                print_seq3(frame, lut, TERM_ESC_COLOR24BIT_FG, r_down, g_down, b_down);
             }
-            printf("\xe2\x96\x84");  // UTF8 bytes of U+2584 (lower half block)
+            bstr_xappend(NULL, frame, bstr0("\xe2\x96\x84"));  // UTF8 bytes of U+2584 (lower half block)
```
Should we `#define` this above like we do for the escape sequences?

kasper93 replied:
done

Traneptora commented on video/out/vo_tct.c:
```diff
@@ -262,17 +248,19 @@ static void flip_page(struct vo *vo)
             p->frame->planes[0], p->frame->stride[0],
             p->opts.term256, p->lut);
     }
+    printf("%.*s\n", BSTR_P(p->frame_buf));
```
Why printf here instead of fwrite? It may be necessary on windows to translate from UTF-8 to the codepage, but if so, shouldn't we have an `#ifdef _WIN32` or something as above?

kasper93 replied:
It is done once per frame, I don't think there is measurable difference.

avih replied:
The `printf` variant still needs to do `strlen` (or something to that effect), because it won't write past the end of the string. The precision value doesn't imply ignoring the terminating null if it arrives before the precision is reached.

If you have the length and it doesn't need any of the things which printf does (like interpret format specifiers, or align, etc), then just [f]write it directly.

kasper93 replied:
> The printf variant still needs to do strlen (or something to that effect), because it won't write past the end of the string. The precision value doesn't imply ignoring the terminating null if it arrives before the precision is reached.

I'm pretty sure lying to printf about string length is basically UB and likely it just blow up and ignore the thing. But sure, no need to call it at all.

avih replied:
> I'm pretty sure lying to printf about string length

You can't lie about the string length because the precision value is not meant to specify the string length. With `%s` and `%S`, it specifies the maximum number of bytes to output (in cases where the string length is more than that).

kasper93 replied:
fixed

Traneptora commented:
Another thought: if we're changing the `#define` escape sequences, does it make more sense to make it a macro or more sense to do `static const bstr foo = bstr0("bar");`?

kasper93 commented:
> Another thought: if we're changing the #define escape sequences, does it make more sense to make it a macro or more sense to do static const bstr foo = bstr0("bar");?

It's not C++, static initializers has to be compile time constant and bstr0 is not.

Traneptora commented:
> > Another thought: if we're changing the #define escape sequences, does it make more sense to make it a macro or more sense to do static const bstr foo = bstr0("bar");?
> 
> It's not C++, static initializers has to be compile time constant and bstr0 is not.

If bstr0 is inlined, would it be a compile-time constant? Or are compilers not that smart?

kasper93 commented:
> If bstr0 is inlined, would it be a compile-time constant? Or are compilers not that smart?

They are smart, but only in C++ where you can mark function as `constexpr`. In C though some builtin functions like `strlen` are const evaluated, so indeed we can try. See https://github.com/mpv-player/mpv/pull/13718/commits/0b07bcc9c987f5cc4649cfd17cd047ca8cbc0772

kasper93 commented:
@na-na-hi: Try with new version, does it still do that? Also use `perf stat` instead of `time`

avih commented:
> Yes. Still slower.

The reason for that is that writing the whole page in one write prevents parallelization between mpv and the terminal.

With small (buffered) writes, the write is typically instant because it's buffered, and after one write the terminal can start processing it, while mpv continues to calculate the next write, and so mpv and the terminal both do processing in parallel.

But with a big write, the terminal is is idle while mpv calculates the output, and then mpv is idle after (during) the write - until the terminal finishes processing the input.

So the throughput can be lower with bigger writes, because mpv and terminal processing are serialized and not parallel.

kasper93 commented:
I've added also fputs for win32, but it got little bit tangled between two PR, so it will go in next batch https://github.com/mpv-player/mpv/commit/fc15e9c83b94d3d4cfc1f00e02f6c3be3524d81e

kasper93 commented on video/out/vo_tct.c:
```diff
@@ -39,15 +39,17 @@
+static const bstr TERM_ESC_COLOR24BIT_BG   = bstr0_s("\033[48;2");
+static const bstr TERM_ESC_COLOR24BIT_FG   = bstr0_s("\033[38;2");
+
+static const bstr UNICODE_LOWER_HALF_BLOCK = bstr0_s("▄");
```
done

kasper93 commented:
> The reason for that is that writing the whole page in one write prevents parallelization between mpv and the terminal.

And I'm fine with that. It is a lot nicer to flip the page "atomically" instead of looking at tearing and cracking of image, because we are writing as we go. 

If it is really performance critical issue, I can do flush after each line. IIRC this also fixed the playback in my terminal. This probably would be good balance, but still flipping frame fully is fast enough for real-time playback.

kasper93 commented:
@na-na-hi 
Can you try, is this makes any difference?
```
diff --git a/video/out/vo_tct.c b/video/out/vo_tct.c
index bc9bd3c670..4ef899b798 100644
--- a/video/out/vo_tct.c
+++ b/video/out/vo_tct.c
@@ -178,6 +178,8 @@ static void write_half_blocks(bstr *frame,
             bstr_xappend(NULL, frame, UNICODE_LOWER_HALF_BLOCK);
         }
         bstr_xappend(NULL, frame, TERM_ESC_CLEAR_COLORS);
+        fwrite(frame->start, frame->len, 1, stdout);
+        frame->len = 0;
     }
 }
```

avih commented:
> It is a lot nicer to flip the page "atomically" instead of looking at tearing and cracking of image, because we are writing as we go.

The former does not necessarily results in the latter.

I.e. the terminal can still process the input in parallel to mpv writing it (in small chunks), but only render it to screen some time later. E.g. on a fixed interval timing, or when there's some idle at the terminal input - and then the terminal decides that it's time to render.

That depends on the terminal. Some terminals do that (render the whole page atomically even with small writes), while some don't.

What is sure though, with all terminals, is that the limit is reached sooner when doing big writes. E.g. frame-drops will definitely happen sooner with full-page-write (e.g. while trying bigger and bigger resolutions, or increasing playback speed more and more).

And the limit will be reached sooner also for terminals which do atomic screen updates even with small writes.

I'm not judging whether that's a good or bad thing, but it is a fact.

kasper93 commented:
> What is sure though, with all terminals, is that the limit is reached sooner when doing big writes. E.g. frame-drops will definitely happen sooner with full-page-write (e.g. while trying bigger and bigger resolutions, or increasing playback speed more and more).

This is wrong, but thank you for your opinion, not a fact.

avih commented:
> That depends on the terminal. Some terminals do that (render the whole page atomically even with small writes), while some don't.

Also, there's an unofficial extension (escape sequences) to handle exactly such cases of big output, which tell the terminal to suspend rendering until some mark (sequence) in the future, so that the application can complete the page/output and then tell the terminal to resume rendering.

This sequence is typically ignored by terminals which don't support its semantics (and it was modified a bit over the years).

See https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036

kasper93 commented:
> See https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036

Interesting, can be added too.

avih commented:
> This is wrong

Thanks for the great explanation and data points, but I don't think it shows how the serialized processing does not make it slower than parallel processing with small writes.

kasper93 commented:
@na-na-hi: I've added `--vo-tct-buffering` so you can tune buffering for your needs. Line buffering works reasonably well, it has still tearing and breaks if anything else is printed in the log, but if it works better for you it is good, we have an option now.

@avih: I've added synchronized output support.

kasper93 commented:
> Thanks. The option works as expected for me.

Thanks for testing. 

Added also removal of `MultiByteToWideChar` if not needed. On my branch https://github.com/mpv-player/mpv/compare/master...kasper93:mpv:puts Again, got little bit tangled, so tct and the other PR would need to go in first, then I can rebase.

kasper93 commented:
Ok, ready to review after recent changes. Particularity I'm curious if `TERM_ESC_SYNC_UPDATE_BEGIN` works fine on your end and don't cause some weirdness in different envs.

kasper93 commented:
I have no more changes to do here. Feel free to play around with it if someone if fancy improving tct :)

N-R-K commented on misc/bstr.h:
```diff
@@ -56,9 +56,11 @@ static inline struct bstr bstrdup(void *talloc_ctx, struct bstr str)
     return r;
 }
 
+#define bstr0_s(s) (struct bstr){(unsigned char *)(s), (s) ? strlen(s) : 0}
```
This isn't portable at all. If you're just trying to initialize a string literal then use `sizeof - 1` which is guaranteed to be a constant expression.

```c
#define bstr0lit(s) { (unsigned char *)(s), sizeof(s) - 1 }
```

kasper93 replied:
I know, but does mpv support any other compiler than clang or gcc? I don't think so and likely other issues would arise if such compilation were tried.

Either way, I can change to sizeof, it is better probably.

N-R-K replied:
> I know, but does mpv support any other compiler than clang or gcc?

Even clang and gcc will not build this if builtin optimizations are disabled (e.g `-fno-builtin`).

Also, one more thing, surrounding the arg with empty string would be a nice improvement since it'd detect if the macro is being used on non string-literals.

```c
// should only be used on string literals
#define bstr0lit(s) { (unsigned char *)("" s ""), sizeof(s) - 1 }
```

kasper93 replied:
ACK, will fix later.

N-R-K replied:
> I added a commit https://github.com/mpv-player/mpv/commit/0a78ed3dca58e57cf85fb57e71cff22e17e9759c to fix this.

Trying to support `bstr0lit(NULL)` is basically zero value. If you want null string, just do `{0}`.

Also, I don't really see any reason to keep the current `bstr0_s` macro around. If someone is looking for optimization/inlining just enable LTO really.
